app/assets/javascripts/libs/diffable_map.js

```diff
@@ -0,0 +1,294 @@
+    // this.id = idCounter++;
+    this.chunks = [];
+    this.existsCache = new Map();
+    this.entryCount = 0;
+    this.itemsPerBatch = itemsPerBatch;
+
+    if (optKeyValueArray != null) {
+      for (const [key, value] of optKeyValueArray) {
+        this.mutableSet(key, value);
+      }
+    }
+  }
+
+  get(key: K): V {
+    let idx = 0;
+    while (this.chunks[idx] != null) {
+      if (this.chunks[idx].has(key)) {
+        // $FlowFixMe
+        return this.chunks[idx].get(key);
+      }
+      idx++;
+    }
+    throw new Error("Get empty");
+  }
+
+  has(key: K): boolean {
+    try {
+      this.get(key);
+      return true;
+    } catch (exception) {
+      return false;
+    }
+  }
+
+  set(key: K, value: V): DiffableMap<K, V> {
+    if (this.existsCache.has(key)) {
+      let idx = 0;
+      while (this.chunks[idx] != null) {
+        if (this.chunks[idx].has(key)) {
+          const newMap = shallowCopy(this);
+          newMap.chunks[idx] = new Map(this.chunks[idx]);
+          newMap.chunks[idx].set(key, value);
+          return newMap;
+        }
+        idx++;
+      }
+      throw new Error("should not happen");
+    } else {
+      const isTooFull = this.entryCount / this.chunks.length > this.itemsPerBatch;
+      const nonFullMapIdx =
+        isTooFull || this.chunks.length === 0 ? -1 : Math.floor(Math.random() * this.chunks.length);
+
+      // Key didn't exist. Add it.
+      const newDiffableMap = shallowCopy(this);
+      newDiffableMap.existsCache.set(key, true);
+      newDiffableMap.entryCount = this.entryCount + 1;
+      if (nonFullMapIdx > -1) {
+        newDiffableMap.chunks[nonFullMapIdx] = new Map(this.chunks[nonFullMapIdx]);
+        newDiffableMap.chunks[nonFullMapIdx].set(key, value);
+        return newDiffableMap;
+      } else {
+        const freshMap = new Map();
+        freshMap.set(key, value);
+        newDiffableMap.chunks.push(freshMap);
+        return newDiffableMap;
+      }
+    }
+  }
+
+  mutableSet(key: K, value: V): void {
+    if (this.existsCache.has(key)) {
+      let idx = 0;
+      while (this.chunks[idx] != null) {
+        if (this.chunks[idx].has(key)) {
+          this.chunks[idx].set(key, value);
+          return;
+        }
+        idx++;
+      }
+    } else {
+      // let idx = 0;
+      const isTooFull = this.entryCount / this.chunks.length > this.itemsPerBatch;
+      const nonFullMapIdx =
+        isTooFull || this.chunks.length === 0 ? -1 : Math.floor(Math.random() * this.chunks.length);
+
+      // Key didn't exist. Add it.
+      this.existsCache.set(key, true);
+      this.entryCount++;
+      if (nonFullMapIdx > -1) {
+        this.chunks[nonFullMapIdx].set(key, value);
+      } else {
+        const freshMap = new Map();
+        freshMap.set(key, value);
+        this.chunks.push(freshMap);
+      }
+    }
+  }
+
+  clone(): DiffableMap<K, V> {
+    const newDiffableMap = new DiffableMap();
```

Comment:
What happens to the `existsCache` and the other properties in this case?

Reply:
Oops, very good point, I'll update this and try to adapt the tests, since they did not catch this.